File: src/ScOutfitter.Core/Ship.cs
using System.Text.Json.Nodes;

namespace ScOutfitter.Core;

/// <summary>One hardpoint the optimizer can fill.</summary>
public sealed class Slot
{
    public required Kind Kind { get; init; }
    public required int Size { get; init; }
    public required string Port { get; init; }
    public string? Equipped { get; init; }
    /// <summary>Gun slot currently carrying a gimbal mount.</summary>
    public bool Gimbal { get; init; }
    /// <summary>Missile rack slots: what the stock rack is loaded with.</summary>
    public string? EquippedMissile { get; init; }
    /// <summary>Bespoke ports (Stingray "Merlin_Nose"): an item must carry these tags.</summary>
    public HashSet<string> RequiredTags { get; init; } = new(StringComparer.Ordinal);
    /// <summary>Tags this port offers to items that require some.</summary>
    public HashSet<string> PortTags { get; init; } = new(StringComparer.Ordinal);
    /// <summary>Port is not editable in game: keep whatever is fitted.</summary>
    public bool Fixed { get; init; }

    public override string ToString() => $"{Kind} S{Size} ({Port})";
}

public sealed class Ship
{
    public required string Name { get; init; }
    public List<Slot> Slots { get; } = [];
    /// <summary>Wiki capacity x 1000; matches quantum_range / fuel_rate.</summary>
    public double QuantumFuelUnits { get; init; }
    public double QuantumRangeM { get; init; }
    public double PowerGeneration { get; init; }
    public double CoolingGeneration { get; init; }

    public IEnumerable<Slot> SlotsOf(Kind kind) => Slots.Where(s => s.Kind == kind);

    public string? DefaultQuantumDrive => SlotsOf(Kind.QuantumDrive).FirstOrDefault()?.Equipped;
}

public static class ShipLoader
{
    private static readonly Dictionary<string, Kind> PortTypes = new(StringComparer.Ordinal)
    {
        ["QuantumDrive"] = Kind.QuantumDrive,
        ["Shield"] = Kind.Shield,
        ["PowerPlant"] = Kind.PowerPlant,
        ["Cooler"] = Kind.Cooler,
        ["Radar"] = Kind.Radar,
    };

    /// <param name="fixedGuns">Replace gimbal mounts by fixed guns of the full hardpoint size.</param>
    /// <param name="mannedTurrets">Include guns inside manned turrets.</param>
    public static Ship FromJson(JsonNode v, bool fixedGuns = true, bool mannedTurrets = false)
    {
        var ship = new Ship
        {
            Name = v.Str("name") is { Length: > 0 } n ? n : "?",
            QuantumFuelUnits = v.Num("quantum", "quantum_fuel_capacity") * 1000,
            QuantumRangeM = v.Num("quantum", "quantum_range"),
            PowerGeneration = v.Num("power", "generation_segments"),
            CoolingGeneration = v.Num("cooling", "generation_segments"),
        };
        HashSet<string> vehicleTags = v.StrSet("port_tags");
        Walk(v.Arr("ports"), ship.Slots, fixedGuns, mannedTurrets, vehicleTags);
        return ship;
    }

    private static bool Compatible(JsonNode? port, string typeName) =>
        port.Arr("compatible_types").Any(c => c.Str("type") == typeName);

    private static (HashSet<string> Required, HashSet<string> Offered) Tags(HashSet<string> vehicleTags, params JsonNode?[] ports)
    {
        var req = new HashSet<string>(StringComparer.Ordinal);
        var offered = new HashSet<string>(vehicleTags, StringComparer.Ordinal);
        foreach (JsonNode? p in ports)
        {
            if (p is null)
            {
                continue;
            }

            req.UnionWith(p.StrSet("required_tags"));
            offered.UnionWith(p.StrSet("port_tags"));
            offered.UnionWith(p.StrSet("required_tags"));
        }

        return (req, offered);
    }

    private static void Walk(IEnumerable<JsonNode> ports, List<Slot> slots, bool fixedGuns, bool manned, HashSet<string> vehicleTags)
    {
        foreach (JsonNode p in ports)
        {
            string type = p.Str("type");
            int size = p.Int("sizes", "max");
            string? equipped = p.StrOrNull("equipped_item", "name");
            List<JsonNode> children = p.Arr("ports").ToList();
            bool editable = p.BoolOr(true, "editable");
            (HashSet<string> req, HashSet<string> offered) = Tags(vehicleTags, p);

            if (type == "MissileLauncher")
            {
                string? missile = children.FirstOrDefault(c => c.Str("type") == "Missile").StrOrNull("equipped_item", "name");
                slots.Add(new Slot
                {
                    Kind = Kind.MissileRack, Size = size, Port = p.Str("name"), Equipped = equipped,
                    EquippedMissile = missile, RequiredTags = req, PortTags = offered, Fixed = !editable,
                });
                continue;
            }

            if (PortTypes.TryGetValue(type, out Kind kind))
            {
                slots.Add(new Slot
                {
                    Kind = kind, Size = size, Port = p.Str("name"), Equipped = equipped,
                    RequiredTags = req, PortTags = offered, Fixed = !editable,
                });
                continue;
            }

            if (Compatible(p, "WeaponGun") && size > 0)
            {
                // a gun hardpoint; may hold a gimbal mount whose child port is one size smaller
                JsonNode? childGun = children.FirstOrDefault(c => c.Str("type") == "WeaponGun" || Compatible(c, "WeaponGun"));
                bool isGimbal = childGun is not null && (equipped ?? string.Empty).Contains("gimbal", StringComparison.OrdinalIgnoreCase);
                string? childEquipped = childGun.StrOrNull("equipped_item", "name");
                (req, offered) = Tags(vehicleTags, p, childGun);

                if (!editable && childGun is not null)
                {
                    // mount is welded on (Stingray): only the gun inside swaps, at the child's size
                    int childSize = childGun.Int("sizes", "max");
                    slots.Add(new Slot
                    {
                        Kind = Kind.Gun, Size = childSize > 0 ? childSize : size, Port = p.Str("name"),
                        Equipped = childEquipped, Gimbal = isGimbal, RequiredTags = req, PortTags = offered,
                        Fixed = !childGun.BoolOr(true, "editable"),
                    });
                }
                else if (fixedGuns || !isGimbal)
                {
                    slots.Add(new Slot
                    {
                        Kind = Kind.Gun, Size = size, Port = p.Str("name"),
                        Equipped = isGimbal ? childEquipped : equipped, Gimbal = false,
                        RequiredTags = req, PortTags = offered, Fixed = !editable,
                    });
                }
                else
                {
                    slots.Add(new Slot
                    {
                        Kind = Kind.Gun, Size = size - 1, Port = p.Str("name"), Equipped = childEquipped, Gimbal = true,
                        RequiredTags = req, PortTags = offered,
                    });
                }

                continue;
            }

            if ((type == "Turret" || type == "TurretBase") && manned)
            {
                Walk(children, slots, fixedGuns, manned, vehicleTags);
            }
        }
    }
}

File: outfitter/ship.py
"""Ship model: which component slots a hull has and what is equipped by default."""
from __future__ import annotations

from dataclasses import dataclass, field

from . import api

PORT_TYPES = {"QuantumDrive": "quantum_drive", "Shield": "shield",
              "PowerPlant": "power_plant", "Cooler": "cooler"}


@dataclass
class Slot:
    kind: str
    size: int
    port: str
    equipped: str | None
    gimbal: bool = False  # gun slot currently carrying a gimbal mount

    def __str__(self) -> str:
        return f"{self.kind} S{self.size} ({self.port})"


@dataclass
class Ship:
    name: str
    slots: list[Slot] = field(default_factory=list)
    quantum_fuel_units: float = 0.0  # wiki capacity * 1000 (matches quantum_range / fuel_rate)
    quantum_range_m: float = 0.0
    power_generation: float = 0.0
    cooling_generation: float = 0.0

    def slots_of(self, kind: str) -> list[Slot]:
        return [s for s in self.slots if s.kind == kind]

    @property
    def default_quantum_drive(self) -> str | None:
        qd = self.slots_of("quantum_drive")
        return qd[0].equipped if qd else None


def _compatible(port: dict, type_name: str) -> bool:
    return any(c.get("type") == type_name for c in (port.get("compatible_types") or []))


def _walk_ports(ports: list[dict], slots: list[Slot], *, fixed_guns: bool, manned: bool) -> None:
    for p in ports or []:
        ptype = p.get("type")
        size = int((p.get("sizes") or {}).get("max") or 0)
        eq = (p.get("equipped_item") or {}).get("name")
        if ptype in PORT_TYPES:
            slots.append(Slot(PORT_TYPES[ptype], size, p["name"], eq))
            continue
        if _compatible(p, "WeaponGun") and size > 0:
            # a gun hardpoint; may currently hold a gimbal mount whose child port is one size smaller
            child_gun = None
            for c in (p.get("ports") or []):
                if c.get("type") == "WeaponGun" or _compatible(c, "WeaponGun"):
                    child_gun = c
            eq_is_gimbal = child_gun is not None and "gimbal" in (eq or "").lower()
            child_eq = (child_gun.get("equipped_item") or {}).get("name") if child_gun else None
            if fixed_guns or not eq_is_gimbal:
                slots.append(Slot("gun", size, p["name"], child_eq if eq_is_gimbal else eq, False))
            else:
                slots.append(Slot("gun", size - 1, p["name"], child_eq, True))
            continue
        if ptype in ("Turret", "TurretBase") and manned:
            _walk_ports(p.get("ports") or [], slots, fixed_guns=fixed_guns, manned=manned)


def load_ship(name: str, *, fixed_guns: bool = True, manned_turrets: bool = False) -> Ship:
    v = api.wiki_vehicle(name)
    ship = Ship(v.get("name") or name)
    q = v.get("quantum") or {}
    ship.quantum_fuel_units = float(q.get("quantum_fuel_capacity") or 0) * 1000
    ship.quantum_range_m = float(q.get("quantum_range") or 0)
    ship.power_generation = float((v.get("power") or {}).get("generation_segments") or 0)
    ship.cooling_generation = float((v.get("cooling") or {}).get("generation_segments") or 0)
    _walk_ports(v.get("ports") or [], ship.slots, fixed_guns=fixed_guns, manned=manned_turrets)
    return ship
